patches/api/0353-WIP-Brain-API.patch

```diff
@@ -0,0 +1,802 @@
++    }
++
++    /**
++     * Gets the miniumum runtime in ticks that this behavior can run.
++     *
++     * @return time in ticks
++     */
++    int getMinRuntime();
++
++    /**
++     * Gets the maximum runtime in ticks that this behavior can run.
++     *
++     * @return time in ticks
++     */
++    int getMaxRuntime();
++
++    /**
++     * Gets if the behavior is still able to run.
++     * If false, this will stop the behavior.
++     *
++     * @return time in ticks
++     */
++    default boolean canStillRun(T entity) {
++        return true;
++    }
++
++    /**
++     * Gets the required memory pairs that this behavior needs inorder to start.
++     *
++     * @return required pairs
++     */
++    Collection<MemoryPair> getMemoryRequirements();
++
++}
+diff --git a/src/main/java/io/papermc/paper/entity/brain/activity/behavior/BehaviorPair.java b/src/main/java/io/papermc/paper/entity/brain/activity/behavior/BehaviorPair.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..5105005291e0f033d615c14fd449298ba758d573
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/brain/activity/behavior/BehaviorPair.java
+@@ -0,0 +1,9 @@
++package io.papermc.paper.entity.brain.activity.behavior;
++
++import org.bukkit.entity.LivingEntity;
++
++/**
++ * Represents a behavior with a priority, used for creating activities.
++ */
++public record BehaviorPair<T extends LivingEntity>(int priority, Behavior<T> behavior) {
++}
+diff --git a/src/main/java/io/papermc/paper/entity/brain/memory/MemoryKeyStatus.java b/src/main/java/io/papermc/paper/entity/brain/memory/MemoryKeyStatus.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..1f7d330d75d0aa76053c3a54786dbfbfac976aa4
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/brain/memory/MemoryKeyStatus.java
+@@ -0,0 +1,19 @@
++package io.papermc.paper.entity.brain.memory;
++
++/**
++ * Represents the current status that a memory may be in.
++ */
++public enum MemoryKeyStatus {
++    /***
++     * Represents a memorykey being only registered.
++     */
++    REGISTERED,
++    /***
++     * Represents a registered memorykey with there currently not being a value present.
++     */
++    ABSENT,
++    /***
++     * Represents a registered memorykey with there being a value present.
++     */
++    PRESENT
++}
+diff --git a/src/main/java/io/papermc/paper/entity/brain/memory/MemoryPair.java b/src/main/java/io/papermc/paper/entity/brain/memory/MemoryPair.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..1ff50af8ac677ae05f60e3deb9a398d2ac9465be
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/brain/memory/MemoryPair.java
+@@ -0,0 +1,9 @@
++package io.papermc.paper.entity.brain.memory;
++
++import org.bukkit.entity.memory.MemoryKey;
++
++/**
++ * A memory pair represents a memory and what key status it is required to have.
++ */
++public record MemoryPair(MemoryKeyStatus requiredStatus, MemoryKey<?> key) {
```

Comment:
I didn't think it was needed because the type of the key isn't really relevant when it's in a memory pair. Typically this is just used with vanilla condition statements. Such as only allowing something to activate when a memory key has reached a certain status.